owncloudApp/src/main/java/com/owncloud/android/ui/activity/PassCodeActivity.kt

```diff
@@ -0,0 +1,425 @@
+     */
+    protected fun clearBoxes() {
+        for (passCodeEditText in mPassCodeEditTexts) {
+            passCodeEditText!!.setText("")
+        }
+        mPassCodeEditTexts[0]!!.requestFocus()
+    }
+
+    /**
+     * Overrides click on the BACK arrow to correctly cancel ACTION_ENABLE or ACTION_DISABLE, while
+     * preventing than ACTION_CHECK may be worked around.
+     *
+     * @param keyCode       Key code of the key that triggered the down event.
+     * @param event         Event triggered.
+     * @return              'True' when the key event was processed by this method.
+     */
+    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
+        if (keyCode == KeyEvent.KEYCODE_BACK && event.repeatCount == 0) {
+            if (ACTION_REQUEST_WITH_RESULT == intent.action || ACTION_CHECK_WITH_RESULT == intent.action) {
+                finish()
+            } // else, do nothing, but report that the key was consumed to stay alive
+            return true
+        }
+        return super.onKeyDown(keyCode, event)
+    }
+
+    /**
+     * Saves the pass code input by the user as the current pass code.
+     */
+    protected fun savePassCodeAndExit() {
+        val resultIntent = Intent()
+        val passCodeString = StringBuilder()
+        for (i in 0 until numberOfPassInputs) {
+            passCodeString.append(mPassCodeDigits[i])
+        }
+        resultIntent.putExtra(KEY_PASSCODE, passCodeString.toString())
+        setResult(RESULT_OK, resultIntent)
+        notifyDocumentProviderRoots(applicationContext)
+        finish()
+    }
+
+    private fun loadPinFromOldFormatIfPossible(): String {
+        val appPrefs = PreferenceManager
+            .getDefaultSharedPreferences(applicationContext)
+
+        var pinString = ""
+        for (i in 1..4)
+            pinString += appPrefs.getString(PREFERENCE_PASSCODE_D + i, null)
+
+        return pinString
+    }
+
+    public override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putBoolean(KEY_CONFIRMING_PASSCODE, mConfirmingPassCode)
+        outState.putStringArray(KEY_PASSCODE_DIGITS, mPassCodeDigits)
+    }
+
+    private inner class PassCodeDigitTextWatcher(private val mIndex: Int, private val mLastOne: Boolean) : TextWatcher {
+        private operator fun next(): Int {
+            return if (mLastOne) 0 else mIndex + 1
+        }
+
+        /**
+         * Performs several actions when the user types a digit in an input field:
+         * - saves the input digit to the state of the activity; this will allow retyping the
+         * pass code to confirm it.
+         * - moves the focus automatically to the next field
+         * - for the last field, triggers the processing of the full pass code
+         *
+         * @param s     Changed text
+         */
+        override fun afterTextChanged(s: Editable) {
+            if (s.length > 0) {
```

Comment:
We can use `isNotEmpty` here 👍 

BTW, it would be great to have more descriptive names here instead of `s`, something like `changedText` or similar, I don't know 🤔